### ISSUE TITLE
Permit user to add directly their social account to an existing account

### ISSUE THREAD

**Issue by sycured** [FusionAuth/fusionauth-issues#751]
### Problem
When an user used one email address different than their email returned by identity provider like Facebook or Google, we have 2 differents users.

##### Classic
User registered using email/password
A day, this user try to connect using Facebook and we've a different account because it's not the same email address.

##### Identity provider
User registered using Facebook
A day, this user try to connect using Google and  we've a different account because it's not the same email address.

### Solution
Permit to an existing user to add identity provider accounts to his actual account (email/password or using another identity provider).

### Alternatives/workarounds
Building a dedicated microservice that reconcile with the existing account inside our database.

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by robotdan** [comment]
Thanks for opening this issue @sycured . This is something we want to handle at some point. This will partially be possible once we deliver multiple identity support via issue https://github.com/FusionAuth/fusionauth-issues/issues/1. 

There are edge cases and difficulties with ensuring user owns these other accounts, but it is something we would like to support.

**Comment by saulfrank** [comment]
After going through this, I think this is what it could look like on the admin.

![image](https://user-images.githubusercontent.com/22662385/89560607-34ca3e80-d80f-11ea-8b14-086e6f1d6355.png)

**Comment by saulfrank** [comment]
@robotdan - I saw there is thinking around multiple emails per account. I actually think that linking social is not necessarily about the email. I would have thought its more to do with the the IDs from oauth. I think having multiple email addresses per account is a separate issue and in my opinion, would be cleaner /safer to link up multiple social ids to one user id.

**Comment by sycured** [comment]
@saulfrank I totally agree, my issue is about linking social account to an existing account created via email/password or another social account.

Having multiple emails on one account can be more like:
- primary email
- emergency email (forgotten access…)
But it's not this issue

**Comment by dakkafex** [comment]
it would be nice if we could only even return the data from the IDP to the client :/

**Comment by robotdan** [comment]
Thanks for the feedback and comments! 

@saulfrank I think we are in agreement. 

Here is where we are going:

1. A user can have one to many identities
2. An identity is a unique Id defined as unique to a particular namespace. 

Up until this point, we just sort of assume that `email` is the unique Id because it is a well understood namespace. 

This has a few problems (probably more than these) :
1. The same user may have one-to-many email addresses
2. Not all applications consider email a unique value. For example, if an application or web site does not perform email validation then I can sign up with your email address and attempt an account takeover. 
3. Not all applications use `email` as a unique Id, they may use `username` (which is not globally unique), a phone number, a unique Id using an integer or encoded string of some sort. 
4. Some unique Ids (email, username, phone number) can change over time and we want to understand when the same "user" has a new identity. 
5. Some systems have an email address, but the user or application may choose not to share it with third parties, so we may not have access to the email address, but a surrogate Id or email address (Sign in with Apple uses this strategy)


Our current thinking is to build namespaces, some we would define, but others could be defined by the user to make this super extensible. This will help us identify a user "identity" as unique within some specified context. 

Example namespaces:

1. Email, this one is fairly easy to understand. `bob@example.com` is globally unique. However, w/out email verification, FusionAuth cannot guarantee that only one real "user" has this email address. 
2. E.164 phone number, unique within the world wide PSTN network. 
3. LDAP `cn`, unique to a particular directory 
4. Facebook unique Id
5. Twitter handles
6. WhatsApp username
7. United States Social Security Number (SSN)
8. IRS Employer Identification Number (EIN)
9. VAT identification number
10. Credit card number
11. Mailing address

Lots more I'm sure, but if we allow these to be defined within FusionAuth, anyone can create their own namespace and then allow this to define an identity type. Once we have an identity type, we can authenticate a user against that type, which means we can take a SSN, Twitter handle, or whatever as the "login Id" on an authentication request. 

This should enable:

1. Linking one to many (social or otherwise) accounts with the same or different email addresses to the same FusionAuth user
2. Allow the change of an email address in Twitter, Facebook, or Google (for example) to not cause a new user to be created in FusionAuth. 
3. Authenticate against an IdP that does not provide an email address to FusionAuth


Issues  that will deliver some of parts of this road map are:
- #1 
- #346 
- #751 (this one)

**Comment by dakkafex** [comment]
well when you login trough FA with idp (github) FA itself gets the data from the github user endpoint, but i havent found anyway to send that object directly back to the client.

react native > oauth login with github > redirects back to react, and all the client gets is an accessToken for a new user that has been created from the authorize enpoint

**Comment by saulfrank** [comment]
This is good news Daniel, I’m excited to see it when it’s available :)

I think it’s right to think of it as an identity platform, with links to
identifiers of identity. I respect all the edge cases to ensure it remains
secure. :)

Best
Saul

**Comment by robotdan** [comment]
@dakkafex 

You will have access to the response in the reconcile lambda, so you can store any or all of that information on the user.
https://fusionauth.io/docs/v1/tech/lambdas/openid-connect-response-reconcile/

If you're asking for FusionAuth to hand you back the `access_token` from GH for you to then call the GH user endpoint directly, we do not currently support that. 

The work around would be for you to make the request to GH, get the access token, and then complete the login with FusionAuth using this JWT with the External JWT IdP. 

https://fusionauth.io/docs/v1/tech/lambdas/external-jwt-reconcile/

This would be `react native > oauth login with github > redirect back to react > send access token to FusionAuth to complete login`. 

If none of these options will work for you, please open a separate GH issue so we can identify the feature, use case, etc and then see if it is something we want to support.
